source/agora/consensus/Validation.d

```diff
@@ -956,6 +957,55 @@ public bool isValidEnrollment (const ref Enrollment enrollment,
     return isInvalidEnrollmentReason(enrollment, block_height, findUTXO) is null;
 }
 
+/*******************************************************************************
+
+    Check the validity of a pre-image information
```

Comment:
You might want to express your documentation in terms of "parent" and "descendant" / "child".
Because there's a clear hierarchy in pre-image: we successively reveal "parent" or "ancestor" of a hash.

Documenting "pre-image" as invalid is a bit confusing to me, because a pre-image is not a self-contained entity, it's defined in terms of its image. What we can test is if a `PreimageInfo` is a parent for another `PreimageInfo`.

Consider the following scenario: You have a `PreimageInfo` at height 98, and you get a `PreimageInfo` for height 90. Except the hash actually matches the pre-image for height 92. The `PreimageInfo` is not a parent for the `PreimageInfo` we have, but the hash itself is a pre-image - just at a different height.

Reply:
The `PreimageInfo` is a model for "pre-image" we described in the white paper. So we can use `PreimageInfo` and "pre-image" interchangingly. Because we are developing a "pre-image" in our product, we can use the word, "pre-image" as `PreimageInfo`. So the "pre-image" gets to have an actual preimage and a height. It's not the pre-image we use in Mathematics or Cryptography. We need to use an word like "pre-image" in our context.
I hope I understood what you mean exactly.

Reply:
> It's not the pre-image we use in Mathematics or Cryptography.

My point was that, since we are borrowing vocabulary from those field, we shouldn't conflate the two. Whenever talking about details, we could use `PreImageInfo`, but in the general sense we refer to them as "preimage".
But the documentation, as it stands, is good enough.

Reply:
> My point was that, since we are borrowing vocabulary from those field, we shouldn't conflate the two. Whenever talking about details, we could use PreImageInfo, but in the general sense we refer to them as "preimage".

Thank your for the additional comment and the merge.